javascript/ql/experimental/adaptivethreatmodeling/test/endpoint_large_scale/ContradictoryEndpointCharacteristics.ql

```diff
@@ -0,0 +1,74 @@
+  EndpointTypes::EndpointType endpointClass, float confidence, string errorMessage
+) {
+  // An endpoint's characteristics should not include positive indicators with medium/high confidence for more than one
+  // class.
+  exists(
+    EndpointCharacteristics::EndpointCharacteristic characteristic2,
+    EndpointTypes::EndpointType endpointClass2, float confidence2
+  |
+    endpointClass.getEncoding() != endpointClass2.getEncoding() and
+    characteristic.appliesToEndpoint(endpoint) and
+    characteristic2.appliesToEndpoint(endpoint) and
+    characteristic.hasImplications(endpointClass, true, confidence) and
+    characteristic2.hasImplications(endpointClass2, true, confidence2) and
+    confidence > characteristic.mediumConfidence() and
+    confidence2 > characteristic2.mediumConfidence() and
+    // We currently know of several high-confidence negative characteristics that apply to some known sinks.
+    // TODO: Experiment with lowering the confidence of `"FileSystemAccess"`, `"DOM"`, `"DatabaseAccess"`, and
+    // `"JQueryArgument"`.
+    not (
+      characteristic = ["TaintedPathSink", "FileSystemAccess"] and
+      characteristic2 = ["TaintedPathSink", "FileSystemAccess"]
+      or
+      characteristic = ["DomBasedXssSink", "DOM"] and
+      characteristic2 = ["DomBasedXssSink", "DOM"]
+      or
+      characteristic = ["DomBasedXssSink", "JQueryArgument"] and
+      characteristic2 = ["DomBasedXssSink", "JQueryArgument"]
+      or
+      characteristic = ["NosqlInjectionSink", "DatabaseAccess"] and
+      characteristic2 = ["NosqlInjectionSink", "DatabaseAccess"]
+      or
+      characteristic = ["SqlInjectionSink", "DatabaseAccess"] and
+      characteristic2 = ["SqlInjectionSink", "DatabaseAccess"]
+    )
+  ) and
+  errorMessage = "Endpoint has high-confidence positive indicators for multiple classes"
+  or
+  // An enpoint's characteristics should not include positive indicators with medium/high confidence for some class and
+  // also include negative indicators with medium/high confidence for this same class.
+  exists(EndpointCharacteristics::EndpointCharacteristic characteristic2, float confidence2 |
+    characteristic.appliesToEndpoint(endpoint) and
+    characteristic2.appliesToEndpoint(endpoint) and
+    characteristic.hasImplications(endpointClass, true, confidence) and
+    characteristic2.hasImplications(endpointClass, false, confidence2) and
+    confidence > characteristic.mediumConfidence() and
+    confidence2 > characteristic2.mediumConfidence()
+  ) and
+  errorMessage = "Endpoint has high-confidence positive and negative indicators for the same class"
+  or
+  // The endpoint's characteristics should not include indicators with confidence outside of [0, 1].
+  characteristic.appliesToEndpoint(endpoint) and
+  characteristic.hasImplications(_, _, confidence) and
+  (confidence < 0 or confidence > 1) and
+  errorMessage = "Endpoint has an indicator with confidence outside of [0, 1]"
```

Comment:
I think this last test case can be phrased independently of the endpoint! You could make it a separate test case/query that looks for only the last 3 lines of this logic, without the `appliesToEndpoint(...)` call.
Slight advantages:
- you will catch discrepancies even if (for whatever reason) the test data doesn't actually contain an endpoint of the given characteristic
- slight speedup in this part of the test

Reply:
Good idea! Done in cf29cde2e8377a0d450267831d343b42b72c8b21